File: python/ipo.py
import heapq


class Solution:
    def findMaximizedCapital(
        self, k: int, w: int, profits: list[int], capital: list[int]
    ) -> int:
        """
        NOTE:
        1. time complexity: O(klogn), space complexity: O(n)
        2. double heap
        """
        min_heap: list[tuple[int, int]] = list(zip(capital, profits))
        heapq.heapify(min_heap)
        max_heap: list[int] = []
        for _ in range(k):
            while len(min_heap) > 0 and w >= min_heap[0][0]:
                _, pro = heapq.heappop(min_heap)
                heapq.heappush(max_heap, -1 * pro)
            if len(max_heap) == 0:
                break
            w += -1 * heapq.heappop(max_heap)
        return w


def test_findMaximizedCapital_case_1():
    """This is a test case for findMaximizedCapital"""
    # arrange
    k: int = 2
    w: int = 0
    profits: list[int] = [1, 2, 3]
    capital: list[int] = [0, 1, 1]
    expected: int = 4

    # act
    solution = Solution()
    actual = solution.findMaximizedCapital(k, w, profits, capital)

    # assert
    assert expected == actual


def test_findMaximizedCapital_case_2():
    """This is a test case for findMaximizedCapital"""
    # arrange
    k: int = 3
    w: int = 0
    profits: list[int] = [1, 2, 3]
    capital: list[int] = [0, 1, 2]
    expected: int = 6

    # act
    solution = Solution()
    actual = solution.findMaximizedCapital(k, w, profits, capital)

    # assert
    assert expected == actual


def test_findMaximizedCapital_case_3():
    """This is a test case for findMaximizedCapital"""
    # arrange
    k: int = 1
    w: int = 0
    profits: list[int] = [1, 2, 3]
    capital: list[int] = [0, 1, 2]
    expected: int = 1

    # act
    solution = Solution()
    actual = solution.findMaximizedCapital(k, w, profits, capital)

    # assert
    assert expected == actual

File: rust/src/code/ipo.rs
struct Solution {}

impl Solution {
    // NOTE:
    // 1. time complexity: O(klogn), space complexity: O(n)
    // 2. double heap
    pub fn find_maximized_capital(k: i32, mut w: i32, profits: Vec<i32>, capital: Vec<i32>) -> i32 {
        use std::cmp::Reverse;
        use std::collections::BinaryHeap;

        let mut min_heap: BinaryHeap<Reverse<(i32, i32)>> = BinaryHeap::new();
        for (pro, cap) in profits
            .into_iter()
            .zip(capital.into_iter())
            .collect::<Vec<(i32, i32)>>()
        {
            min_heap.push(Reverse((cap, pro)));
        }
        let mut max_heap: BinaryHeap<i32> = BinaryHeap::new();
        for _ in 0..k {
            while min_heap
                .peek()
                .filter(|Reverse((min_w, _))| w >= *min_w)
                .is_some()
            {
                if let Some(Reverse((_, pro))) = min_heap.pop() {
                    max_heap.push(pro);
                }
            }
            if let Some(p) = max_heap.pop() {
                w += p;
            } else {
                break;
            }
        }
        w
    }
}

#[cfg(test)]
mod test {
    use super::Solution;

    #[test]
    fn test_find_maximized_capital_case_1() {
        // arrange
        let k: i32 = 2;
        let w: i32 = 0;
        let profits: Vec<i32> = vec![1, 2, 3];
        let capital: Vec<i32> = vec![0, 1, 1];
        let expected: i32 = 4;

        // act
        let actual = Solution::find_maximized_capital(k, w, profits, capital);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_find_maximized_capital_case_2() {
        // arrange
        let k: i32 = 3;
        let w: i32 = 0;
        let profits: Vec<i32> = vec![1, 2, 3];
        let capital: Vec<i32> = vec![0, 1, 2];
        let expected: i32 = 6;

        // act
        let actual = Solution::find_maximized_capital(k, w, profits, capital);

        // assert
        assert_eq!(expected, actual);
    }
}
